src/client/dfs/dfs.c

```diff
@@ -468,11 +470,9 @@ fetch_entry(daos_handle_t oh, daos_handle_t th, const char *name, size_t len,
 		sgl->sg_nr_out	= 0;
 		sgl->sg_iovs	= sg_iovs;
 
-		rc = daos_obj_fetch(oh, th, 0, &dkey, 1, iod, sgl, NULL,
-				    NULL);
+		rc = daos_obj_fetch(oh, th, DAOS_COND_DKEY_FETCH, &dkey, 1, iod, sgl, NULL, NULL);
 		if (rc) {
-			D_ERROR("Failed to fetch entry %s "DF_RC"\n", name,
-				DP_RC(rc));
+			D_ERROR("Failed to fetch entry %s "DF_RC"\n", name, DP_RC(rc));
```

Comment:
so what if this returns -DER_NONEXIST?   Still an error?

Reply:
yes at this point we are not checking for existence anymore as the first fetch should have done that.
this is the second fetch of the same dkey but this time fetching the symlink value only assuming the dkey is exist.
if it doesn't then something weird happen like a punch in between, which should be an error, or when using balanced mode with DTX, i assume this will be restarted.

Reply:
this applies only to symlink values, so remove_dir_contents does not apply IMO.
exists value is internal, so also for user, doesn't matter. we can actually remove this exists flag since we switch to conditional i believe, but i will leave that to a different PR.
so let's take the case you are mentioning, non-dtx, and someone punched in between. that means truly the entry does not exist anymore. so why would we not return this error if the user has done this race case?

Reply:
why remove_dir_contents() does not apply? remove_dir_contents() is used to implement rmdir by force that may contains symlink child entry. I do not think "-DER_NONEXIST" is the expected errno for rmdir since the dir itself (that is being removed) exists.

Reply:
ok i see what you are saying. yes that makes sense. probably better to remove this exists flag here then and check rc value in this case. will fix

Reply:
ok i updated this, but your comment on remove_dir_contents actually still does not apply. because when we fetch the entry set explicitly not to fetch the symlink value, so we never enter this anyway